File: sdk/typescript/src/peer.ts
/**
 * AIP peer: send helpers with automatic per-stream sequence assignment.
 */
import {
  ErrorCode,
  Message,
  MessageOptions,
  ResultStatus,
  makeAction,
  makeError,
  makeEvent,
  makeResult,
  messageFromDict,
} from "./messages.js";
import { Receiver, Sequencer } from "./sequence.js";
import { Session } from "./session.js";
import { Transport } from "./transport.js";

export class AIPPeer {
  readonly source: string;
  readonly sessionId: string;
  readonly identity: string;
  transport?: Transport;
  seq = new Sequencer();
  receiver = new Receiver();
  session: Session;

  constructor(
    source: string,
    sessionId: string,
    options: { transport?: Transport; identity?: string } = {},
  ) {
    this.source = source;
    this.sessionId = sessionId;
    this.identity = options.identity ?? source;
    this.transport = options.transport;
    this.session = new Session(sessionId);
    this.session.receiver = this.receiver;
  }

  send(m: Message): Message {
    if (!m.seq) m.seq = this.seq.next(this.sessionId, this.source);
    this.transport?.send(m);
    return m;
  }

  sendEvent(name: string, data?: Record<string, unknown>, opts: MessageOptions = {}): Message {
    return this.send(makeEvent(this.sessionId, this.source, name, data, opts));
  }

  sendAction(
    name: string,
    options: {
      target?: string;
      params?: Record<string, unknown>;
      expect?: { event: string };
    } = {},
    opts: MessageOptions = {},
  ): Message {
    return this.send(makeAction(this.sessionId, this.source, name, options, opts));
  }

  sendResult(
    status: ResultStatus,
    inReplyTo: string,
    options: {
      data?: Record<string, unknown>;
      code?: string;
      message?: string;
      duplicate?: boolean;
      originalResultId?: string;
      retry?: { allowed: boolean; after_ms?: number };
    } = {},
    opts: MessageOptions = {},
  ): Message {
    return this.send(makeResult(this.sessionId, this.source, status, inReplyTo, options, opts));
  }

  sendError(
    code: ErrorCode,
    options: { message?: string; retry?: { allowed: boolean; after_ms?: number } } = {},
    opts: MessageOptions = {},
  ): Message {
    return this.send(makeError(this.sessionId, this.source, code, options, opts));
  }

  handle(raw: unknown): { kind: "accept" | "duplicate" | "stale" | "gap"; message: Message } {
    const msg = messageFromDict(raw as Record<string, unknown>);
    const kind = this.receiver.classify(msg);
    if (kind === "accept") this.receiver.apply(msg);
    return { kind, message: msg };
  }
}

File: sdk/python/aip/peer.py
"""AIP peer: send helpers with automatic per-stream sequence assignment.

A peer is one participant (decision engine or executor) on one side of a
session. It assigns seq per (session, source), builds messages, and can
classify inbound messages against its own receiver (SPEC §5.4).
"""
from typing import Optional

from .messages import (
    Message,
    make_action,
    make_error,
    make_event,
    make_result,
)
from .sequence import Receiver, Sequencer
from .session import Session


class AIPPeer:
    def __init__(
        self,
        source: str,
        session_id: str,
        transport=None,
        identity: Optional[str] = None,
    ) -> None:
        self.source = source
        self.session_id = session_id
        self.transport = transport
        self.identity = identity or source
        self.seq = Sequencer()
        self.receiver = Receiver()
        self.session = Session(session_id)
        self.session.receiver = self.receiver

    # --- outbound -----------------------------------------------------------
    def send(self, message: Message) -> Message:
        if not message.seq:
            message.seq = self.seq.next(self.session_id, self.source)
        if self.transport is not None:
            self.transport.send(message.to_dict())
        return message

    def send_event(self, name, data=None, **kw) -> Message:
        return self.send(make_event(self.session_id, self.source, name, data, **kw))

    def send_action(self, name, target=None, params=None, expect=None, **kw) -> Message:
        return self.send(
            make_action(self.session_id, self.source, name, target, params, expect, **kw)
        )

    def build_action(self, name, target=None, params=None, expect=None, **kw) -> Message:
        """Build an action WITHOUT sending it. Useful when the caller must
        register correlation state before delivery (a synchronous transport
        may re-enter the caller before send() returns)."""
        return make_action(self.session_id, self.source, name, target, params, expect, **kw)

    def send_result(
        self,
        status,
        in_reply_to,
        *,
        data=None,
        code=None,
        message=None,
        duplicate=False,
        original_result_id=None,
        retry=None,
        **kw,
    ) -> Message:
        return self.send(
            make_result(
                self.session_id, self.source, status, in_reply_to,
                data=data, code=code, message=message,
                duplicate=duplicate, original_result_id=original_result_id,
                retry=retry, **kw,
            )
        )

    def send_error(self, code, *, message=None, retry=None, in_reply_to=None, **kw) -> Message:
        return self.send(
            make_error(self.session_id, self.source, code, message=message,
                       retry=retry, in_reply_to=in_reply_to, **kw)
        )

    # --- inbound ------------------------------------------------------------
    def handle(self, raw: dict) -> tuple:
        """Classify an inbound message.

        Returns (classification, message) where classification is one of
        "accept", "duplicate", "stale", "gap". The caller applies side
        effects for "accept" only; "duplicate" must never be re-executed.
        """
        msg = Message.from_dict(raw)
        cls = self.receiver.classify(msg)
        if cls == "accept":
            self.receiver.apply(msg)
        return cls, msg